lib/Monads/wp/WP_Pre.thy

```diff
@@ -51,21 +50,20 @@ fun pre_tac trace ctxt pre_rules used_thms_ref i t = let
     then Seq.empty else Seq.single t2 end
     handle Option => Seq.empty
 
-fun tac trace used_thms_ref ctxt = let
-    val pres = Named_Theorems.get ctxt @{named_theorems wp_pre}
-  in pre_tac trace ctxt pres used_thms_ref end
-
 val method =
   let
     val used_thms_ref = Unsynchronized.ref [] : (string * string * term) list Unsynchronized.ref
   in
-    Args.context >> (fn _ => fn ctxt =>
-      Method.SIMPLE_METHOD' (tac (Config.get ctxt wp_trace) used_thms_ref ctxt))
+    Attrib.thms >> (fn thms => fn ctxt =>
+      Method.SIMPLE_METHOD' (pre_tac (Config.get ctxt wp_trace) ctxt thms used_thms_ref))
   end
 end
 \<close>
 
-method_setup wp_pre0 = \<open>WP_Pre.method\<close>
+method_setup pre_tac = \<open>WP_Pre.method\<close>
+
+named_theorems wp_pre
+method wp_pre0 declares wp_pre = pre_tac wp_pre
 method wp_pre = wp_pre0?
```

Comment:
Speaking only for myself, I find having the same name for the method and named thms a bit confusing. It probably is the right way to go in terms of actual use, and only confusing during implementation.

Reply:
It is a bit confusing in implementation, but I think it is better for the user, because there are fewer names to remember. It's also an old tradition for things like `simp`. It'd probably have been better back then (25 years ago or so) to introduce the convention of `simp_thms`, but I don't think we should do our own one now.